sdk/zksend/src/channel/index.ts

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { safeParse } from 'valibot';
+
+import { withResolvers } from '../utils/withResolvers.js';
+import type { ZkSendRequestTypes, ZkSendResponsePayload, ZkSendResponseTypes } from './events.js';
+import { ZkSendResponse } from './events.js';
+
+const DEFAULT_ZKSEND_ORIGIN = 'https://zksend.com';
+
+export class ZkSendPopup {
+	#id: string;
+	#origin: string;
+	#close?: () => void;
+
+	constructor(origin = DEFAULT_ZKSEND_ORIGIN) {
+		this.#id = crypto.randomUUID();
+		this.#origin = origin;
+	}
+
+	async createRequest<T extends keyof ZkSendResponseTypes>(
+		type: T,
+		data: ZkSendRequestTypes[T],
+	): Promise<ZkSendResponseTypes[T]> {
+		const { promise, resolve, reject } = withResolvers<ZkSendResponseTypes[T]>();
+
+		let popup: Window | null = null;
+
+		const listener = (event: MessageEvent) => {
+			if (event.origin !== this.#origin) {
+				return;
+			}
+			const { success, output } = safeParse(ZkSendResponse, event.data);
+			if (!success || output.id !== this.#id) return;
+
+			window.removeEventListener('message', listener);
+
+			if (output.payload.type === 'reject') {
+				reject(new Error('TODO: Better error message'));
+			} else if (output.payload.type === 'resolve') {
+				resolve(output.payload.data as ZkSendResponseTypes[T]);
+			}
+		};
+
+		this.#close = () => {
+			popup?.close();
+			window.removeEventListener('message', listener);
+		};
+
+		window.addEventListener('message', listener);
+
+		popup = window.open(
+			`${origin}/dapp/${type}?${new URLSearchParams({
+				id: this.#id,
+				origin: this.#origin,
```

Comment:
Can you also add a `name` or `title` prop that is supposed to be the name of the dapp? I realized I missed this but it's in the designs.